torch/jit/annotations.py

```diff
@@ -51,50 +51,92 @@ def __getitem__(self, name):
 
-def get_signature(fn, rcb, loc):
+def get_signature(fn, rcb, loc, is_method):
     # Python 3.5 adds support for the nice annotation syntax, so try that first.
+    signature = None
     if PY35:
-        sig = try_real_annotations(fn)
-        if sig is not None:
-            return sig
+        signature = try_real_annotations(fn)
+        if signature is not None and is_method:
+            # If this is a method, then the signaure will include a type for
+            # `self`, but type comments do not contain a `self`. So strip it
+            # away here so everything is consistent (`inspect.ismethod` does
+            # not work here since `fn` is unbound at this point)
+            param_types, return_type = signature
+            param_types = param_types[1:]
+            signature = (param_types, return_type)
+
+    if signature is None:
+        type_line, source = None, None
+        try:
+            source = dedent(''.join(get_source_lines_and_file(fn)[0]))
+            type_line = get_type_line(source)
+        except TypeError:
+            pass
+        # This might happen both because we failed to get the source of fn, or
+        # because it didn't have any annotations.
+        if type_line is not None:
+            signature = parse_type_line(type_line, rcb, loc)
+
+    return signature
+
+
+def is_function_or_method(the_callable):
+    # A stricter version of `inspect.isroutine` that does not pass for built-in
+    # functions
+    return inspect.isfunction(the_callable) or inspect.ismethod(the_callable)
+
+
+def is_vararg(the_callable):
+    if not is_function_or_method(the_callable) and hasattr(the_callable, '__call__'):  # noqa: B004
+        # If `the_callable` is a class, de-sugar the call so we can still get
+        # the signature
+        the_callable = the_callable.__call__
+
+    if is_function_or_method(the_callable):
+        if PY2:
+            # [inspect args]
+            # `inspect.getfullargspec` is not available in Python 2 but
+            # `inspect.getargspec` is deprecated in Python 3, so we have to
+            # switch over them
+            return inspect.getargspec(the_callable).varargs is not None
+        else:
+            return inspect.getfullargspec(the_callable).varargs is not None
+    else:
+        return False
 
-    type_line, source = None, None
-    try:
-        source = dedent(''.join(get_source_lines_and_file(fn)[0]))
-        type_line = get_type_line(source)
-    except TypeError:
-        pass
-    # This might happen both because we failed to get the source of fn, or
-    # because it didn't have any annotations.
-    if type_line is None:
-        return None
 
-    return parse_type_line(type_line, rcb, loc)
+def get_param_names(fn, n_args):
+    if not is_function_or_method(fn) and hasattr(fn, '__call__') and is_function_or_method(fn.__call__):  # noqa: B004
+        # De-sugar calls to classes
+        fn = fn.__call__
 
+    if is_function_or_method(fn):
```

Comment:
This should always return true no? we maybe should just throw if at this point it's not a function or method.